pkg/daemon/daemon.go

```diff
@@ -634,16 +637,16 @@ func (dn *Daemon) Run(stopCh <-chan struct{}, exitCh <-chan error) error {
 
 	go wait.Until(dn.worker, time.Second, stopCh)
 
-	for {
-		select {
-		case <-stopCh:
-			return nil
-		case <-signaled:
-			return nil
-		case err := <-exitCh:
-			// This channel gets errors from auxiliary goroutines like loginmonitor and kubehealth
-			glog.Warningf("Got an error from auxiliary tools: %v", err)
-		}
+	select {
+	case <-stopCh:
+		return nil
+	case sig := <-signaled:
+		glog.Warningf("shutdown of machine-config-daemon triggered via signal %d", sig)
+		return fmt.Errorf("shutdown of the machine-config-daemon trigger via syscall signal %d", sig)
```

Comment:
I think getting `SIGTERM` isn't an error, it's a normal condition.  Which relates to one of the original goals I had here in that in the "idle" aka "not applying updates" case, we don't install a `SIGTERM` handler at all - we just let the kernel unilaterally kill the process.

I more recently posted about this here: https://internals.rust-lang.org/t/should-rust-programs-unwind-on-sigint/13800/11

See also e.g. https://github.com/nhorman/rng-tools/pull/72

And related to all this, ideally of course we do https://github.com/openshift/machine-config-operator/issues/1190 - then we shouldn't need to handle SIGTERM at all in the MCO.  With that if we're killed (or the machine power cycled/kernel froze) in the middle of an update, we either have the old or new system.

Reply:
I did go back on and forth on whether or to report `SIGTERM` as an error or not for the reasons your blog highlights and ultimately went with returning the error to signal the reason for the death of the process (and thus make it obvious in the logs). In retrospect, I think just logging and moving on is the better path. 

FWIW I was arguing for making the MCD block reboots, however, @crawford argued that fully transactional updates would negate the need and is the better decision. My view is that until we get the transactional update mechanism, a short-lived inhibitor is _better_ than nothing and could prevent some support cases and bug reports. 

> And related to all this, ideally of course we do #1190 - then we shouldn't need to handle SIGTERM at all in the MCO. 

I would disagree -- we should at, the very least, through away the pending transaction (or have `rpm-ostree` do it).